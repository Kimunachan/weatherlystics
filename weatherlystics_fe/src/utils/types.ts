import { z } from "zod";
import { weatherDataSchema } from "./schemas";

export type WeatherDataType = z.infer<typeof weatherDataSchema>;
