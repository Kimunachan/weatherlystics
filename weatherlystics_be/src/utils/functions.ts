export const range = (start: number, stop: number, step: number = 1) =>
  Array.from({ length: (stop - start) / step }, (_, i) => start + i * step);
