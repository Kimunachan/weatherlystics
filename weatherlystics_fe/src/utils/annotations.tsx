import { ChartDataset, PluginOptionsByType } from 'chart.js';

    type MinMax = {
        min: number;
        max: number;
        xMin: number | string;
        xMax: number | string;
    };
    
    type DatasetItem = {
        x: number | string;
        y: number;
    }

    export const getMinMaxPerDataset = (datasets: ChartDataset<'line'>[]): MinMax[] => {
        return datasets.map(dataset => {
        const values = dataset.data.map(item => (typeof item === 'object' ? (item as DatasetItem).y : item)) as number[];
        const min = Math.min(...values);
        const max = Math.max(...values);
        const minIndex = values.indexOf(min);
        const maxIndex = values.indexOf(max);
        const xMin = typeof dataset.data[minIndex] === 'object' ? (dataset.data[minIndex] as DatasetItem).x : minIndex;
        const xMax = typeof dataset.data[maxIndex] === 'object' ? (dataset.data[maxIndex] as DatasetItem).x : maxIndex;
        return { min, max, xMin, xMax };
        });
    };

    export const createAnnotations = (minMax: MinMax[]): PluginOptionsByType<'line'>['annotation']['annotations'] => {
        return minMax.flatMap((range) => [
        {
            type: 'point' as const,
            xValue: range.xMin,
            yValue: range.min,
            backgroundColor: 'red',
            radius: 5,
            label: {
            content: `Min: ${range.min}`,
            position: 'start' as const,
            enabled: true,
            },
        },
        {
            type: 'point' as const,
            xValue: range.xMax,
            yValue: range.max,
            backgroundColor: 'blue',
            radius: 5,
            label: {
            content: `Max: ${range.max}`,
            position: 'start' as const,
            enabled: true,
            },
        },
        ]);
    };