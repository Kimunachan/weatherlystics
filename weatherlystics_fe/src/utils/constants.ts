export const customStyles = {
    control: (provided) => ({
      ...provided,
      height: 35,
      minHeight: 30,
      width: 150,
      fontSize: 12,
      borderRadius: '15px',
      borderColor: '#fff',
      backgroundColor: '#121212',
      color: '#fff',
      '&:hover': {
        borderColor: '#ccc',
      },
    }),
    singleValue: (provided) => ({
      ...provided,
      color: '#fff',
    }),
    input: (provided) => ({
      ...provided,
      color: '#fff',
    }),
    dropdownIndicator: (provided) => ({
      ...provided,
      color: '#fff',
    }),
    clearIndicator: (provided) => ({
      ...provided,
      color: '#fff',
    }),
    menu: (provided) => ({
      ...provided,
      backgroundColor: '#121212',
      color: '#fff',
      fontSize: 13,
      width: 250,
      
      // Verstecken Sie die Scrollbar
      overflow: 'auto',
      scrollbarWidth: 'none', // Für Firefox
      '&::-webkit-scrollbar': {
        display: 'none', // Für Chrome, Safari und Opera
      },
    }),
    option: (provided, state) => ({
      ...provided,
      color: state.isSelected ? '#121212' : '#fff',
      backgroundColor: state.isSelected ? '#fff' : '#121212',
      '&:hover': {
        color: '#121212',
        backgroundColor: '#ccc',
      },
    }),
  };