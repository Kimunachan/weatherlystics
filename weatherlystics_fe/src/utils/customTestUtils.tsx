import ReactQueryProvider from "@/components/ReactQueryProvider";
import {
  RenderResult,
  render as customRender,
  renderHook as customRenderHook,
} from "@testing-library/react";
import React from "react";

const render = (ui: JSX.Element, { ...renderOptions } = {}): RenderResult => {
  const wrapper = ({ children }: { children?: React.ReactNode }) => {
    return <ReactQueryProvider>{children}</ReactQueryProvider>;
  };
  return customRender(ui, { wrapper, ...renderOptions });
};

const renderHook = <TResult, TProps>(callback: (props: TProps) => TResult) =>
  customRenderHook<TResult, TProps>(callback, {
    wrapper: ({ children }) => (
      <ReactQueryProvider>{children}</ReactQueryProvider>
    ),
  });

export * from "@testing-library/react";

export { render, renderHook };
