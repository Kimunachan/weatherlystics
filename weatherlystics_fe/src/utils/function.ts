import { WeatherDataType } from "./types";

export const reformData = (data: WeatherDataType) => {
  const timeZones = data.timezone;

  const hourly = data.hourly.map((data) => {
    return {
      time: getHours(data.time, timeZones),
      temperature: data.temperature2m,
      humidity: data.relativeHumidity2m,
      apparentTemperature: data.apparentTemperature,
    };
  });

  const hourLabels = hourly.map((data) => data.time);
  const temperatureData = hourly.map((data) => data.temperature);
  const humidityData = hourly.map((data) => data.humidity);
  const apparentTemperatureData = hourly.map(
    (data) => data.apparentTemperature
  );

  const hourlyData = {
    labels: hourLabels,
    datasets: [
      {
        label: "Temperature",
        data: temperatureData,
        borderColor: "rgba(255, 99, 132, 1)",
        tension: 0.1,
        yAxisID: "y-axis-temp",
      },
      {
        label: "Humidity",
        data: humidityData,
        borderColor: "rgba(54, 162, 235, 1)",
        tension: 0.1,
        yAxisID: "y-axis-humid",
      },
      {
        label: "Apparent Temperature",
        data: apparentTemperatureData,
        borderColor: "rgba(75, 192, 192, 1)",
        tension: 0.1,
        yAxisID: "y-axis-temp",
      },
    ],
  };

  // const dailyData = data.daily.map((data) => {
  //     return {
  //     time: getDay(data.time, timeZones),
  //     weatherCode: data.weatherCode,
  //     temperatureMax: data.temperature2mMax,
  //     temperatureMin: data.temperature2mMin,
  //     apparentTemperatureMax: data.apparentTemperatureMax,
  //     apparentTemperatureMin: data.apparentTemperatureMin,
  //     };
  // });

  return { hourlyData };
};

export const getHours = (time: Date, timezone: string) => {
  const date = new Date(time.toString());
  return date.toLocaleTimeString("de-DE", {
    hour: "2-digit",
    minute: "2-digit",
    timeZone: timezone,
  });
};
