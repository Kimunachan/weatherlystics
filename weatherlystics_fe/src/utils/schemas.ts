import { z } from "zod";

export const weatherDataSchema = z.object({
  latitude: z.number(),
  longitude: z.number(),
  timezone: z.string(),
  timezoneAbbreviation: z.string(),
  current: z.object({
    time: z.date(),
    temperature2m: z.number(),
    relativeHumidity2m: z.number(),
    apparentTemperature: z.number(),
  }),
  hourly: z.array(
    z.object({
      time: z.date(),
      temperature2m: z.number(),
      relativeHumidity2m: z.number(),
      apparentTemperature: z.number(),
    })
  ),
  daily: z.array(
    z.object({
      time: z.date(),
      weatherCode: z.number(),
      temperature2mMax: z.number(),
      temperature2mMin: z.number(),
      apparentTemperatureMax: z.number(),
      apparentTemperatureMin: z.number(),
    })
  ),
});

export const schema = z.object({
  lat: z
    .number()
    .min(-90, "Latitude needs to be at least -90")
    .max(90, "Latitude cant be greater than 90"),
  long: z
    .number()
    .min(-180, "Longitude needs to be at least -180")
    .max(180, "Longitude cant be greater than 180"),
  timezone: z.string(),
  date: z.date(),
  secondDate: z.date().optional(),
});
export const rowSchema = z.object({
  lat: z.number(),
  long: z.number(),
  timezone: z.string(),
  date: z.string(), // using string for the date input value
});

// Define the form schema
export const formSchema = z.object({
  rows: z.array(rowSchema),
});