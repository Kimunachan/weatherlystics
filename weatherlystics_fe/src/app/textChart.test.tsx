import { render, screen, waitFor } from "@/utils/customTestUtils";
import { reformData } from "@/utils/function";
import { WeatherDataType } from "@/utils/types";
import TextChart from "./textChart";

jest.mock('../utils/function', () => ({
  reformData: jest.fn()
}));

const mockData: WeatherDataType = {
  latitude: 0,
  longitude: 0,
  timezone: "",
  timezoneAbbreviation: "",
  current: {
    time: new Date(),
    temperature2m: 20,
    relativeHumidity2m: 50,
    apparentTemperature: 22,
  },
  hourly: [],
  daily: [],
};

const mockReformData = {
  temperatureChart: {
    labels: ["10:00", "11:00"],
    datasets: [{ data: [20, 21] }]
  },
  humidityChart: {
    labels: ["10:00", "11:00"],
    datasets: [{ data: [50, 55] }]
  },
  apparentTemperatureChart: {
    labels: ["10:00", "11:00"],
    datasets: [{ data: [22, 23] }]
  }
};

describe("TextChart", () => {
  beforeEach(() => {
    (reformData as jest.Mock).mockImplementation(() => mockReformData);
  });

  it("renders without crashing", async () => {
    const { container } = render(<TextChart weatherData={mockData} />);

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });

    expect(screen.getByTestId("chart_temp")).toBeInTheDocument();
    expect(screen.getByTestId("chart_humidity")).toBeInTheDocument();
    expect(screen.getByTestId("chart_appTemp")).toBeInTheDocument();
  });

  it("returns null when weatherData is undefined", async () => {
    const { container } = render(<TextChart weatherData={undefined} />);

    await waitFor(() => {
      expect(container).toBeEmptyDOMElement();
    });
  });
  

  it("renders the text data correctly", async () => {
    render(<TextChart weatherData={mockData} />);

    await waitFor(() => {
      expect(screen.getByTestId("chart_temp")).toBeInTheDocument();
      expect(screen.getByTestId("chart_humidity")).toBeInTheDocument();
      expect(screen.getByTestId("chart_appTemp")).toBeInTheDocument();
    });

    expect(screen.getByText("10:00: 20 째C")).toBeInTheDocument();
    expect(screen.getByText("11:00: 21 째C")).toBeInTheDocument();
    expect(screen.getByText("10:00: 50 %")).toBeInTheDocument();
    expect(screen.getByText("11:00: 55 %")).toBeInTheDocument();
    expect(screen.getByText("10:00: 22 째C")).toBeInTheDocument();
    expect(screen.getByText("11:00: 23 째C")).toBeInTheDocument();
  });

  it("renders 'data is not available' when chart data is null", async () => {
    const mockEmptyData = {
      temperatureChart: null,
      humidityChart: null,
      apparentTemperatureChart: null
    };
    
    (reformData as jest.Mock).mockImplementation(() => mockEmptyData);

    render(<TextChart weatherData={mockData} />);

    await waitFor(() => {
      expect(screen.getAllByText("Data is not available")).toBeInTheDocument();
    });
  });
});
