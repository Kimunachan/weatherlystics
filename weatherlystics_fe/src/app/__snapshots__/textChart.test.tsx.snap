// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextChart renders without crashing 1`] = `
<div>
  <div
    class="text_container"
  >
    <section
      class="chart_temp"
      data-testid="chart_temp"
    >
      <div
        class="formRow"
      >
        <div>
          <h3 />
          <ul>
            <li>
              10:00
              : 
              20
               
              째C
            </li>
            <li>
              11:00
              : 
              21
               
              째C
            </li>
          </ul>
        </div>
      </div>
    </section>
    <section
      class="chart_humidity"
      data-testid="chart_humidity"
    >
      <div
        class="formRow"
      >
        <div>
          <h3 />
          <ul>
            <li>
              10:00
              : 
              50
               
              %
            </li>
            <li>
              11:00
              : 
              55
               
              %
            </li>
          </ul>
        </div>
      </div>
    </section>
    <section
      class="chart_appTemp"
      data-testid="chart_appTemp"
    >
      <div
        class="formRow"
      >
        <div>
          <h3 />
          <ul>
            <li>
              10:00
              : 
              22
               
              째C
            </li>
            <li>
              11:00
              : 
              23
               
              째C
            </li>
          </ul>
        </div>
      </div>
    </section>
  </div>
</div>
`;
